src/storage/src/hummock/store/version.rs

```diff
@@ -295,63 +289,33 @@ impl HummockReadVersion {
+                    let mut check_consecutive = false;
+                    for data in &self.staging.data {
+                        if let Some(imm) = data.to_imm()
+                            && sst_imm_ids.contains(&imm.batch_id())
+                        {
+                            assert!(intersect_imm_ids.is_empty() || check_consecutive);
+                            intersect_imm_ids.insert(imm.batch_id());
+                            check_consecutive = true;
+                        } else {
+                            check_consecutive = false;
+                        }
+                    }
 
                     if !intersect_imm_ids.is_empty() {
                         // Check 2)
-                        debug_assert!(check_subset_preserve_order(
-                            intersect_imm_ids.iter().copied(),
-                            self.staging.imm.iter().map(|imm| imm.batch_id()).rev(),
-                        ));
-
-                        // Check 3) and replace imms with a staging sst
-                        for imm_id in &intersect_imm_ids {
-                            if let Some(imm) = self.staging.imm.back() {
-                                if *imm_id == imm.batch_id() {
-                                    self.staging.imm.pop_back();
-                                }
-                            } else {
-                                let local_imm_ids = self
-                                    .staging
-                                    .imm
-                                    .iter()
-                                    .map(|imm| imm.batch_id())
-                                    .collect_vec();
-
-                                unreachable!(
-                                    "should not reach here staging_sst.size {},
-                                    staging_sst.imm_ids {:?},
-                                    staging_sst.epochs {:?},
-                                    local_imm_ids {:?},
-                                    intersect_imm_ids {:?}",
-                                    staging_sst_ref.imm_size,
-                                    staging_sst_ref.imm_ids,
-                                    staging_sst_ref.epochs,
-                                    local_imm_ids,
-                                    intersect_imm_ids,
-                                );
+                        self.staging.data.retain(|data| match data {
+                            StagingData::ImmMem(imm) => {
+                                !intersect_imm_ids.contains(&imm.batch_id())
                             }
-                        }
-                        self.staging.sst.push_front(staging_sst_ref);
+                            StagingData::Sst(_) => true,
+                        });
+                        self.staging.data.push(StagingData::Sst(staging_sst_ref));
+                        self.staging.data.sort_by_key(|data| data.min_epoch());
```

Comment:
If we have more than one imm/sst belong to the same epoch, will sorting by epoch breaks the correct ordering? I still think sorting by epoch is a bit risky. Given that we have found the consecutive IMMs to be replaced, isn't it more straight-forward to replace these IMMs with the staging SST instead of pushing staging sst to the end and sorting by min epoch?

Reply:
@wenym1  Since that we add offset to tail of epoch, could we assume that every memtable own different epoch?

Reply:
Yes, every memtable of a single local state store instance owns different epochs. But I'm afraid that we cannot use the min_epoch of sst here.

Let's use `epoch<a>-<b>` to represent an `EpochWithGap` with real epoch to be `a` and inner offset as `b`, and we have two instances (`A` and `B`) of local state store.

In `epoch1`, `A` writes an imm with epoch1-1, and then a spill happened and generated a SST with min epoch as epoch1-1. And then `B` writes an imm with epoch1-1, and then `A` writes an imm with epoch1-2, and then a spill happened and generated a SST with min epoch also as epoch1-1. Both ssts contains the data of `A`, so they will both appear in `A`, but they have the same min epoch as `epoch1-1`, and this will cause some order problem if we simply sort by min epoch.

I think https://github.com/risingwavelabs/risingwave/pull/15832#discussion_r1570305805 can solve the problem in a more elegant way, and we can avoid the sorting to reduce the time.

Reply:
Anyway, I refactor this code and replace the memtable with staing-data rather than sort all array.

I think refactor it with `BTreeMap` can left in next PR